mlx/ops.cpp

```diff
@@ -2758,4 +2760,91 @@ array dequantize(
   return w_full;
 }
 
+std::map<char, int> string_to_char_map(std::string inp) {
+  std::map<char, int> counts;
+  for (int i = 0; i < inp.size(); i++) {
+    counts[inp[i]] = i;
+  }
+  return counts;
+}
+
+array einsum(
+    std::string equation,
+    const std::vector<array>& operands,
+    StreamOrDevice s /* = {} */) {
+  std::string output;
+  std::string input = equation;
+  if (equation.find("->") == std::string::npos) {
+    output = std::string(equation);
+    std::sort(output.begin(), output.end());
+  } else {
+    auto pos = equation.find("->");
+    output = equation.substr(pos + 2);
+    input = equation.substr(0, pos);
+  }
+  std::vector<std::string> inputs;
+  std::stringstream ss(input);
+  std::string token;
+  while (getline(ss, token, ',')) {
+    inputs.push_back(token);
+  }
+  if (operands.size() != inputs.size()) {
+    throw std::runtime_error(
+        "Number of operands must match the number of input characters");
+  }
+
+  std::map<char, int> input_map;
+  for (int i = 0; i < inputs.size(); i++) {
+    auto arr = operands[i];
+    auto inp = inputs[i];
+    for (int j = 0; j < MIN(arr.shape().size(), inp.size()); j++) {
+      input_map[inp[j]] = arr.shape().at(j);
+    }
+  }
+  std::vector<int> broad;
+  for (auto key : input_map) {
+    broad.push_back(key.second);
+  }
+  std::vector<array> inputs_arr;
+  for (int i = 0; i < operands.size(); i++) {
+    auto arr = operands[i];
+    auto ord_map = string_to_char_map(inputs[i]);
+    std::vector<int> new_shape;
+    for (auto key : input_map) {
+      if (ord_map.find(key.first) != ord_map.end()) {
+        new_shape.push_back(key.second);
+      } else {
+        new_shape.push_back(1);
+      }
+    }
+    std::vector<int> axis;
+    for (auto key : ord_map) {
+      axis.push_back(key.second);
+    }
+    inputs_arr.push_back(
+        broadcast_to(reshape(transpose(arr, axis, s), new_shape, s), broad, s));
+  }
+
+  auto ord_output = string_to_char_map(output);
+  std::vector<int> rhs_order;
+  for (auto key : ord_output) {
+    rhs_order.push_back(key.second);
+  }
+
+  std::vector<int> sum_axis;
+  int i = 0;
+  for (auto key : input_map) {
+    if (ord_output.find(key.first) == ord_output.end()) {
+      sum_axis.push_back(i);
+    }
+    i += 1;
+  }
+
+  auto acc = ones_like(inputs_arr.at(0), s);
+  for (auto arr : inputs_arr) {
+    acc = multiply(acc, arr, s);
+  }
+  return transpose(sum(acc, sum_axis, false, s), rhs_order, s);
```

Comment:
In some/many common cases we should dispatch to `matmul` rather than `multiply` and `sum` as it will be a lot slower and memory inefficient. Is it possible to include that logic (or some of it) now?